crates/primitives/src/receipt.rs

```diff
@@ -16,37 +16,77 @@ pub struct Receipt {
     pub success: bool,
     /// Gas used
     pub cumulative_gas_used: u64,
-    /// Bloom filter.
-    pub bloom: Bloom,
     /// Log send from contracts.
     pub logs: Vec<Log>,
 }
 
 impl Receipt {
+    /// Cacluates [`Log`]'s bloom filter. this is slow operatio and [ReceiptWithBloom] can
+    /// be used to cache this value.
+    pub fn bloom_slow(&self) -> Bloom {
+        logs_bloom(self.logs.iter())
+    }
+}
+
+impl From<Receipt> for ReceiptWithBloom {
+    fn from(receipt: Receipt) -> Self {
+        let bloom = receipt.bloom_slow();
+        ReceiptWithBloom { receipt, bloom }
+    }
+}
+
+/// [`Receipt`] with calculated bloom filter.
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+#[main_codec]
+pub struct ReceiptWithBloom {
+    /// Bloom filter build from logs.
+    pub bloom: Bloom,
+    /// Main receipt body
+    pub receipt: Receipt,
+}
+
+impl ReceiptWithBloom {
+    /// Create new [ReceiptWithBloom]
+    pub fn new(receipt: Receipt, bloom: Bloom) -> Self {
+        Self { receipt, bloom }
+    }
+
+    /// Desolve structure and return receipt
+    pub fn into_receipt(self) -> Receipt {
+        self.receipt
+    }
+
+    /// Desolve structure and return its components
```

Comment:
```suggestion
    /// Consume the structure, returning the receipt and the bloom filter
```

Consume seems to be the standard language I think